### PR TITLE
Project adt7420

### PR DESCRIPTION
**Pull request by ranechita**
Added iio support for adt7420 and example projects

### CODE REVIEW

**Comment by ranechita** [comment]
v2: 

- fixed doxygen errors

File: drivers/temperature/adt7420/iio_adt7420.c [hunk 1]
```diff
@@ -0,0 +1,530 @@
...
+		return -ENOMEM;
+
+	ret = adt7420_init(&desc->adt7420_dev, *(init_param->adt7420_dev_init));
+	if (ret) {
+		free(desc);
+		return ret;
+	}
+	ret = adt7420_reset(desc->adt7420_dev);
+	if (ret) {
+		free(desc);
```
Comment (buha) [inline drivers/temperature/adt7420/iio_adt7420.c:157]: I think if you exit the function here you also need to call:

```
/* Free the resources allocated by adt7420_init(). */
int32_t adt7420_remove(struct adt7420_dev *dev);
```

File: drivers/temperature/adt7420/iio_adt7420.c [hunk 2]
```diff
@@ -0,0 +1,530 @@
...
+	struct adt7420_dev *adt7420;
+
+	uint16_t temp = 0;
+	int32_t temp_c = 0;
+	iio_adt7420 = (struct adt7420_iio_dev *) dev;
+	adt7420 = iio_adt7420->adt7420_dev;
+
+	switch (channel->address) {
+	case 0:
+		if (adt7420_is_spi(adt7420))
```
Comment (buha) [inline drivers/temperature/adt7420/iio_adt7420.c:283]: Shouldn't this interface logic be handled in the device driver itself ? 
IIO part of the driver seems like an odd place to do this to me.
Obviously this implies some small changes to the driver itself.

**Review by amiclaus — COMMENTED** [review]
looks good overall, some comments on my side. Also, adding a *.json file should be included in the PR so the project build is performed.

File: drivers/temperature/adt7420/iio_adt7420.c (hunk 2 above)
Comment (buha) [reply drivers/temperature/adt7420/iio_adt7420.c:283]: ok, this family is a bit odd, it has different way of addressing the registers on spi and on i2c, so i guess this is fine

File: drivers/temperature/adt7420/iio_adt7420.c [hunk 3]
```diff
@@ -0,0 +1,530 @@
...
+			adt7420_reg_read(adt7420, ADT7320_REG_T_HIGH, &temp);
+		else {
+			uint16_t temp_msb = 0, temp_lsb = 0;
```
[the remark this answers is not in the document]
Comment (ranechita) [reply drivers/temperature/adt7420/iio_adt7420.c:289]: If the device is spi, the variables would be declared and then unused. Should they be moved at the start of the function regardless?

File: projects/adt7420-pmdz/src/examples/dummy/dummy_example.c [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+	float temp_c_max, temp_c_min;
+	int ret;
+
+	ret = adt7420_init(&adt7420, adt7420_user_init);
+	if (ret)
+		goto error;
+	ret = adt7420_reset(adt7420);
+	if (ret)
+		goto error;
+	no_os_mdelay(300);
```
[the remark this answers is not in the document]
Comment (ranechita) [reply projects/adt7420-pmdz/src/examples/dummy/dummy_example.c]: yes, it is the standard value for device reset. The delay was now moved in the driver implementation of adt7420_reset function

**Comment by ranechita** [comment]
v3:

- added builds.json file for project build
- added a new commit for the driver files with macros for calculating the temperature value
- modify driver adt7420_get_temperature function to use macros 
- use same macros for calculating temperature in iio functions
- addressed nits

File: projects/adt7420-pmdz/src/platform/maxim/main.c [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+#if (IIO_EXAMPLE+DUMMY_EXAMPLE != 1)
+#error Selected example projects cannot be enabled at the same time. \
+Please enable only one example and re-build the project.
+#endif
+
+error:
+	return 0;
+}
+
```
Comment (amiclaus) [inline projects/adt7420-pmdz/src/platform/maxim/main.c:90]: nit: extra line

File: projects/adt7420-pmdz/src/platform/maxim/main.c [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+#ifdef DUMMY_EXAMPLE
+	struct no_os_uart_desc *uart;
+
+	ret = no_os_uart_init(&uart, &uip);
+	if (ret)
+		goto error;
+
+	maxim_uart_stdio(uart);
+	ret = dummy_example_main();
+	if (ret < 0)
```
Comment (amiclaus) [inline projects/adt7420-pmdz/src/platform/maxim/main.c:78]: `if (ret)` ?

File: projects/adt7420-pmdz/src/platform/maxim/main.c [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+ *
+ * @return ret - Result of the enabled examples execution.
+*******************************************************************************/
+int main ()
+{
+	int ret;
+
+#ifdef IIO_EXAMPLE
+	ret = iio_example_main ();
+	if (ret < 0)
```
[the remark this answers is not in the document]
Comment (amiclaus) [reply projects/adt7420-pmdz/src/platform/maxim/main.c:65]: this doesn't seem to be solved/ at least replied.

**Review by amiclaus — COMMENTED** [review]
some small comments that need to be addressed. also we need to check if the project build finishes successfully prior the merge.

File: projects/adt7420-pmdz/src/platform/maxim/main.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (ranechita) [reply projects/adt7420-pmdz/src/platform/maxim/main.c]: I must've mixed it up with the other chages and forgit about it, my bad. It's fixed now

**Comment by ranechita** [comment]
v4:

- fixed missing/exra spaces, indentations
- addressed styling issues
- reduced sleep after device reset

File: projects/adt7420-pmdz/src/examples/dummy/dummy_example.c [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+#include "dummy_example.h"
+#include "common_data.h"
+#include "adt7420.h"
+#include "no_os_delay.h"
+#include "no_os_print_log.h"
+
+/******************************************************************************/
+/********************** Macros and Constants Definitions **********************/
+/******************************************************************************/
+#define ADT7320_L8				NO_OS_BIT(8)
```
[the remark this answers is not in the document]
Comment (amiclaus) [reply projects/adt7420-pmdz/src/examples/dummy/dummy_example.c]: this does not seem to be solved. check the raw file to see the actual indentations:

image

File: projects/adt7420-pmdz/src/examples/dummy/dummy_example.c [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+#include "common_data.h"
+#include "adt7420.h"
+#include "no_os_delay.h"
+#include "no_os_print_log.h"
+
+/******************************************************************************/
+/********************** Macros and Constants Definitions **********************/
+/******************************************************************************/
+#define ADT7320_L8			NO_OS_BIT(8)
+#define ADT7320_L16			NO_OS_BIT(16)
```
Comment (amiclaus) [inline projects/adt7420-pmdz/src/examples/dummy/dummy_example.c:53]: inconsistent indentation with the rest of the defines.

File: projects/adt7420-pmdz/src/platform/maxim/main.c [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+#ifdef DUMMY_EXAMPLE
+#include "dummy_example.h"
+#endif
+
+/***************************************************************************//**
+ * @brief Main function execution for MAX32655 platform.
+ *
+ * @return ret - Result of the enabled examples execution.
+*******************************************************************************/
+int main ()
```
Comment (amiclaus) [inline projects/adt7420-pmdz/src/platform/maxim/main.c:59]: nit: `int main()`

File: projects/adt7420-pmdz/src/platform/maxim/parameters.h [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+#include "i2c_extra.h"
+#include "irq_extra.h"
+#include "gpio_extra.h"
+
+/******************************************************************************/
+/********************** Macros and Constants Definitions **********************/
+/******************************************************************************/
+
+#ifdef IIO_SUPPORT
+#define INTC_DEVICE_ID 0
```
Comment (amiclaus) [inline projects/adt7420-pmdz/src/platform/maxim/parameters.h:57]: please properly indent all these defines.

File: projects/adt7420-pmdz/src/examples/dummy/dummy_example.c [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+	float temp_c_max, temp_c_min;
+	float temp_now;
+	int ret;
+
+	ret = adt7420_init(&adt7420, adt7420_user_init);
+	if (ret)
+		goto error;
+	ret = adt7420_reset(adt7420);
+	if (ret)
+		goto error;
```
Comment (amiclaus) [inline projects/adt7420-pmdz/src/examples/dummy/dummy_example.c:86]: mem leak, we should call `adt7420_remove` starting from this conditional jump

File: projects/adt7420-pmdz/src/platform/maxim/main.c [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+	struct no_os_uart_desc *uart;
+
+	ret = no_os_uart_init(&uart, &uip);
+	if (ret)
+		goto error;
+
+	maxim_uart_stdio(uart);
+	ret = dummy_example_main();
+	if (ret)
+		goto error;
```
Comment (amiclaus) [inline projects/adt7420-pmdz/src/platform/maxim/main.c:79]: mem leak: `no_os_uart_remove` should be called starting from this conditional jump.

**Review by amiclaus — APPROVED** [review]
looks good, some possible memory leaks and minor indenting issues should be fixed prior the merge.

File: projects/adt7420-pmdz/src/examples/dummy/dummy_example.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (ranechita) [reply projects/adt7420-pmdz/src/examples/dummy/dummy_example.c]: You're right, visual studio seems to display the indentation a little weird

**Comment by ranechita** [comment]
v5:

- fix indentation 
- fix mem leaks

**Pull request merged**